mem-ruby: Implement no_alloc Far Atomics in CHI

vsoriap opened the pull request:
This PR introduces a missing pice of far atomic implementation.  This pull request incorporates several changes:

- Enable 2-level and 4-level (and N-level) cache hierarchies, removing Atomic_NoWait transactions
- Fix Unique Near policy implementation that raised abort
- Add support for alloc_on_atomic == False. Enables Far Atomics on systems where the HNF does not allocate evicted lines at LLC (Like in WriteUpdate).

vsoriap committed:
mem-ruby: Remove AtomicReturn_NoWait from CHI

To make Atomic transaction recursive and enable 2-level config,
remove AtomicReturn_NoWait and other level-dependent code

GitHub Issue: https://github.com/gem5/gem5/issues/882

Change-Id: Iac468cdb8a3b5914c8f05c5cedde866ce85f359a

vsoriap committed:
mem-ruby: Fix policy_type condition in CHI

Fix if-else condition in CHI-cache-actions to correctly
support policy_type Present Near (2)

Change-Id: Ib776d847a908a8ac7693c2d10405bc0c4a9d767d

vsoriap committed:
mem-ruby: Implement Atomic No Alloc Policy

Add alternative implementation to far atomics when the flag alloc_on_commit
is false. The implementation fetches the data, performs the atomic and
writes back the cache line to main memory.

Co-authored-by: Fabian Schätzle <f.schaetzle@fz-juelich.de>
Change-Id: I8797fbc68448e1866a292f4afeedd3613113dddd

vsoriap commented:
This PR should fix the problems experienced by @yichungchen84 in the issue #882

ivanaamit commented:
@tiagormk, can you please review this PR? It fixes the issue #882. Thanks.

yichungchen84 commented:
@vsoriap Thanks, I will have a quick check over the weekend.

fschaetzle commented:
@vsoriap Thanks for the solution. I run two different configurations:
1. core clusters with priv. L1/L2, shared L3 and HN without SLC.
2. cores with priv. L1/L2 and HN incl. SLC.

By adding RUSC state to AtomicReturn and AtomicNoReturn transition the first config run successfully.

However, the second config run into assert(tbe.expected_req_resp.hasExpected()) from Receive_RespSepDataFromCompData action. Looks like the assertion is thrown by L2. After receiving DBIDResp it also receives CompData_I from HN before sending NCBWrData. Maybe a race condition?

vsoriap commented:
Hi, @fschaetzle. 

Could you describe a little bit more the second config. I would like to know how many CPUs and the cache controller flags you set in that config (e.g., alloc_on_atomic, policy_type, allow_SD, etc)

fschaetzle commented:
@vsoriap I tested small configurations with 4 CPUs, allow_SD=true, policy_type=1 and only HN allocates on atomic.
I think [this](https://github.com/fschaetzle/gem5/commit/7bf08d73d5b1995f888bbd40ef7123b697e0e986) is a possible solution for the issues I encountered (at least for small systems and I will verify bigger systems over the weekend).
It adds:
- RUSC states in AtomicReturn and AtomicNoReturn transition.
- CheckCacheFill action
- HN (alloc_on_atomic=true) waits for NCBWrData before sending CompData. Otherwise we run into the assertion mentioned above.

vsoriap commented:
You are right, @fschaetzle !! I accidentally removed the CheckCacheFill and forgot to remove the NoBlock clause in the CompData. I've updated the latest change to add those changes and added you as a co-author.

I've also changed how the WriteFE/BE latency is defined to only have one of each per transaction, as in WriteUpdate.

fschaetzle commented:
@vsoriap thank you. From my side everything is fine for the PR.

ivanaamit commented:
@tiagormk, could you please review this PR? The changes are related to the CHI protocol. Thank you.

tiagormk reviewed (approved):
LGTM! Thanks @vsoriap for the fixes and @fschaetzle for testing.
